inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/MqttReader.java

```diff
@@ -0,0 +1,222 @@
+        options.setMqttVersion(mqttVersion);
+
+        try {
+            synchronized (MqttReader.class) {
+                client = new MqttClient(serverURI, clientId, new MemoryPersistence());
+                client.setCallback(new MqttCallback() {
+                    @Override
+                    public void connectionLost(Throwable cause) {
+                        LOGGER.info("the mqtt connection is lost, try to reconnect");
+                        reconnect();
+                    }
+
+                    @Override
+                    public void messageArrived(String topic, MqttMessage message) throws Exception {
+                        Map<String, String> headerMap = new HashMap<>();
+                        headerMap.put("record.topic", topic);
+                        headerMap.put("record.messageId", String.valueOf(message.getId()));
+                        headerMap.put("record.qos", String.valueOf(message.getQos()));
+                        byte[] recordValue = message.getPayload();
+                        mqttMessagesQueue.put(new DefaultMessage(recordValue, headerMap));
+
+                        LOGGER.debug("the mqtt receive message: {}", new String(recordValue));
+
+                        readerMetric.pluginReadSuccessCount.incrementAndGet();
+                        readerMetric.pluginReadCount.incrementAndGet();
+                    }
+
+                    @Override
+                    public void deliveryComplete(IMqttDeliveryToken token) {
+                    }
+                });
+                client.connect(options);
+                client.subscribe(topic, 1);
+            }
+            LOGGER.info("the mqtt subscribe topic is [{}], qos is [{}]", topic, qos);
+        } catch (Exception e) {
+            LOGGER.error("init mqtt client error ", e);
+        }
+    }
+
+    private void reconnect() {
+        if (!client.isConnected()) {
+            try {
+                client.connect(options);
+                LOGGER.info("the mqtt client reconnect success");
+            } catch (MqttSecurityException e) {
+                LOGGER.error("reconnect mqtt client error ", e);
+            } catch (MqttException e) {
+                LOGGER.error("reconnect mqtt client error ", e);
+            }
+        }
+    }
+
+    @Override
+    public Message read() {
+        if (!mqttMessagesQueue.isEmpty()) {
+            return getMqttMessage();
+        } else {
+            return null;
+        }
+    }
+
+    private DefaultMessage getMqttMessage() {
+        return mqttMessagesQueue.poll();
+    }
+
+    @Override
+    public boolean isFinished() {
+        return finished;
+    }
+
+    @Override
+    public String getReadSource() {
+        return instanceId;
+    }
+
+    @Override
+    public void setReadTimeout(long mill) {
+    }
+
+    @Override
+    public void setWaitMillisecond(long millis) {
+    }
+
+    @Override
+    public String getSnapshot() {
```

Comment:
Could you please complete the strategy of snapshot?

Reply:
@pocozh   when the mqtt topic is subscribed, the message is pushed to the client through the broker. There is no offset attribute, and the snapshot may not be necessary.